### PR TITLE
Feature/dev 265/eip900 staking interface

### PR DESCRIPTION
**Pull request by tavakyan**
Adding basic structure for two validator sets: reserved (privileged set) and staking set. The staking set uses an EIP900 implementation that keeps track of the stake of stakers.  It uses a standard burnable token.  Also added a mock contract to be used in testing the gelt engine modification to accept a smart-contract.

### CODE REVIEW

File: contracts/StakingSet.sol [hunk 1]
```diff
@@ -0,0 +1,11 @@
+pragma solidity ^0.4.23;
+
+import "./interfaces/ValidatorSet.sol";
+import "./StakeBank.sol";
+
+contract StakingSet is ValidatorSet {
+  StakeBank stakeBank;
+
+  constructor() public {
```
Comment (davesag) [inline contracts/StakingSet.sol]: if the constructor doesn't do anything you don't need it here

File: contracts/StakeBank.sol [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        return stakes[stakes.length-1].amount;
+    }
+
+    /// @notice Returns total tokens staked.
+    /// @return amount of tokens staked.
+    function totalStaked() public view returns (uint256) {
+        return totalStakedAt(block.number);
+    }
+
+    /// @notice Returns if history related functions are implemented.
```
Comment (davesag) [inline contracts/StakeBank.sol]: missing the word 'true', as in 'Returns true if history...'

Comment (davesag) [reply contracts/StakeBank.sol]: could just say 'Are history related functions implemented?'

File: contracts/StakeBank.sol [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    /// @notice Returns the total tokens staked at block.
+    /// @param blockNumber Block number to check.
+    /// @return amount of tokens staked.
+    function totalStakedAt(uint256 blockNumber) public view returns (uint256) {
+        return stakedAt(stakeHistory, blockNumber);
+    }
+
+    function updateCheckpointAtNow(Checkpoint[] storage history, uint256 amount, bool isUnstake) internal {
```
Comment (davesag) [inline contracts/StakeBank.sol]: missing comments — for consistency comment every function.

File: contracts/StakeBank.sol [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        Checkpoint storage checkpoint = history[length];
+
+        if (isUnstake) {
+            checkpoint.amount = checkpoint.amount.sub(amount);
+        } else {
+            checkpoint.amount = checkpoint.amount.add(amount);
+        }
+    }
+
+    function stakedAt(Checkpoint[] storage history, uint256 blockNumber) internal view returns (uint256) {
```
Comment (davesag) [inline contracts/StakeBank.sol]: missing comments

File: contracts/mocks/GeltFixedValidatorSet.sol [hunk 5]
```diff
@@ -0,0 +1,48 @@
+pragma solidity ^0.4.23;
+
+contract GeltFixedValidatorSet {
+  // Current list of addresses entitled to participate in the consensus.
+  address[] public validatorsList = [
+    0x0082a978b3f5962a5b0957d9ee9eef472ee55b42f1,
+    0x007d577a597b2742b498cb5cf0c26cdcd726d39e6e
+  ];
+
+  constructor() public {
```
Comment (davesag) [inline contracts/mocks/GeltFixedValidatorSet.sol]: remove unused constructor

File: contracts/mocks/GeltFixedValidatorSet.sol [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  /// the "change" finalized is the activation of the initial set.
+  function finalizeChange() public {
+
+  }
+
+  // Reporting functions: operate on current validator set.
+  // malicious behavior requires proof, which will vary by engine.
+
+  function reportBenign(address validator, uint256 blockNumber) public {
+
```
Comment (davesag) [inline contracts/mocks/GeltFixedValidatorSet.sol]: maybe insert a comment explicitly stating why this is an empty function and use `solhint` meta comments to flag it

File: contracts/mocks/GeltFixedValidatorSet.sol [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  function reportMalicious(address validator, uint256 blockNumber, bytes proof) public {
+
```
Comment (davesag) [inline contracts/mocks/GeltFixedValidatorSet.sol]: maybe insert a comment explicitly stating why this is an empty function and use `solhint` meta comments to flag it

File: contracts/ReserveSet.sol [hunk 8]
```diff
@@ -0,0 +1,12 @@
+pragma solidity ^0.4.23;
+
+import "./MajoritySet.sol";
+
+import "./StakingSet.sol";
+
+contract ReserveSet is MajoritySet {
+  StakingSet stakingSet;
```
Comment (davesag) [inline contracts/ReserveSet.sol]: how is this accessed?

File: contracts/StakingSet.sol [hunk 9]
```diff
@@ -0,0 +1,8 @@
+pragma solidity ^0.4.23;
+
+import "./interfaces/ValidatorSet.sol";
+import "./StakeBank.sol";
+
+contract StakingSet is ValidatorSet {
+  StakeBank stakeBank;
```
Comment (davesag) [inline contracts/StakingSet.sol]: how is this accessed?

Comment (tavakyan) [reply contracts/StakingSet.sol]: Hey, yes a lot of this is work in progress. I have some ideas on how to restructure, but I wanted to get some updates in and see if I can prepare it more for the build automation part

**Pull request merged**